File: ruby/12-god-object/test/account_test.rb
# frozen_string_literal: true

require 'minitest/autorun'
require_relative '../src/account'

class AccountTest < Minitest::Test
  def setup
    @account = Account.new('a@b.com', 'secret')
  end

  def test_logs_in_with_correct_password
    assert_equal true, @account.login('secret')
  end

  def test_rejects_wrong_password
    assert_equal false, @account.login('wrong')
  end

  def test_logs_out
    assert_equal true, @account.logout
  end

  def test_updates_profile
    result = @account.update_profile('Alice', 'Developer')
    assert_equal 'Alice', result[:name]
    assert_equal 'Developer', result[:bio]
  end

  def test_changes_password
    assert_equal true, @account.change_password('secret', 'new')
    assert_equal true, @account.login('new')
  end

  def test_rejects_bad_old_password
    assert_equal false, @account.change_password('wrong', 'new')
  end

  def test_adds_payment_method
    count = @account.add_payment_method('Visa-1234')
    assert_equal 1, count
  end

  def test_removes_payment_method
    @account.add_payment_method('Visa-1234')
    result = @account.remove_payment_method(0)
    assert_equal [], result
  end

  def test_sets_notification_preference
    result = @account.set_notification_preference(:sms, true)
    assert_equal true, result[:sms]
  end

  def test_exports_all_data
    result = @account.export_data
    assert_equal 'a@b.com', result[:email]
    assert result[:audit_log]
  end

  def test_logs_access
    count = @account.log_access('view')
    assert_equal 1, count
  end

  def test_checks_subscription
    assert_equal 'basic', @account.check_subscription
  end

  def test_upgrades_subscription
    result = @account.upgrade_subscription('pro')
    assert_equal 'pro', result
  end
end

File: csharp/12-god-object/test/account_test.cs
using Xunit;

public class AccountTest
{
    [Fact]
    public void LogsInWithCorrectPassword()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.True(acc.login("secret"));
    }

    [Fact]
    public void RejectsWrongPassword()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.False(acc.login("wrong"));
    }

    [Fact]
    public void LogsOut()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.True(acc.logout());
    }

    [Fact]
    public void UpdatesProfile()
    {
        var acc = new Account("a@b.com", "secret");
        var result = acc.update_profile("Alice", "Developer");
        Assert.Equal("Alice", result["name"]);
        Assert.Equal("Developer", result["bio"]);
    }

    [Fact]
    public void ChangesPassword()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.True(acc.change_password("secret", "new"));
        Assert.True(acc.login("new"));
    }

    [Fact]
    public void RejectsBadOldPassword()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.False(acc.change_password("wrong", "new"));
    }

    [Fact]
    public void AddsPaymentMethod()
    {
        var acc = new Account("a@b.com", "secret");
        var count = acc.add_payment_method("Visa-1234");
        Assert.Equal(1, count);
    }

    [Fact]
    public void RemovesPaymentMethod()
    {
        var acc = new Account("a@b.com", "secret");
        acc.add_payment_method("Visa-1234");
        var result = acc.remove_payment_method(0);
        Assert.Empty(result);
    }

    [Fact]
    public void SetsNotificationPreference()
    {
        var acc = new Account("a@b.com", "secret");
        var result = acc.set_notification_preference("sms", true);
        Assert.True(result["sms"]);
    }

    [Fact]
    public void ExportsAllData()
    {
        var acc = new Account("a@b.com", "secret");
        var result = acc.export_data();
        Assert.Equal("a@b.com", result.email);
        Assert.NotNull(result.auditLog);
    }

    [Fact]
    public void LogsAccess()
    {
        var acc = new Account("a@b.com", "secret");
        var count = acc.log_access("view");
        Assert.Equal(1, count);
    }

    [Fact]
    public void ChecksSubscription()
    {
        var acc = new Account("a@b.com", "secret");
        Assert.Equal("basic", acc.check_subscription());
    }

    [Fact]
    public void UpgradesSubscription()
    {
        var acc = new Account("a@b.com", "secret");
        var result = acc.upgrade_subscription("pro");
        Assert.Equal("pro", result);
    }
}
